lnst/RecipeCommon/Perf/Measurements/Results/FlowMeasurementResults.py

```diff
@@ -84,33 +84,11 @@ def end_timestamp(self):
 
     @property
     def warmup_end(self):
-        if self.warmup_duration == 0:
-            return self.start_timestamp
-
-        return max(
-            [
-                parallel[self.warmup_duration - 1].end_timestamp
-                for parallel in (
-                    *self.generator_results,
-                    *self.receiver_results,
-                )
-            ]
-        )
+        return self.start_timestamp+self.warmup_duration
```

Comment:
I think this is not correct.

The original code was basically returing the end_timestamp of the 3rd sample of the results of the **last started** measurement.

Now you return the start_timestamp (whatever) of the **first started** measurement. The `self.start_timestamp` returns `min(measurements)`, so it may be way off what we expect I think.

Reply:
i wasn't trying to keep the old logic, IMO the old logic is the incorrect one because it uses the `warmup duration` (which is in seconds) as an index into an "arbitrary" list... which may or may not be samples (that may or may not be 1 second long)

this is why i think this should be fixed...

but yes, this changes the logic and the new results may be off from the old ones... I don't think it will be common, i think for most cases the difference will be minimal. But we definitely need to check

Reply:
Well consider this situation of parallel results:
```
                    samples

    measurement1    x   x   x   x   x   x   x   x
    measurement2      x   x   x   x   x   x   x   x
    measurement3        x   x   x   x   x   x   x
    measurement4          x   x   x   x   x   x   x

    time (s) -->    1   2   3   4   5   6   7
                    |        |
                    '--------'
                        |
                      warmup
```

With your implementation, the measurement4 would not be completely warmed up,
as we would take the measurement1.start_timestamp as the basis for the addition.

All I'm saying is that we can't use `self.start_timestamp` but most likely
`max([res.start_timestamp) for res in self.generator_results + self.receiver_results + ...])`.

Reply:
The reason was that we originally wanted to completely remove warm{up,down} samples

that doesn't work either... the samples are only APPROXIMATELY 1s long even if we set sampling at exactly 1s... so this assumption is simply wrong.

    Your approach could possibly include part of warm{up,down} sample in sliced results.

so could the old one

    But I believe, that should be fine as no other parts of LNST expects results to have the exact number of samples.

we do additional alignment of measurements for multiple measurements (cpu vs flow measurement) as well to align everything properly... and that will also create "partial" samples... So actually... working with an EXACT number of samples is incorrect everywhere in the code that we currently have.

Besides... even if we do a 60s measurement with 1s samples we can still end up with 61 samples... even when doing no warmup/alignment etc...

Reply:
> Well consider this situation of parallel results:
> 
> ```
>                     samples
> 
>     measurement1    x   x   x   x   x   x   x   x
>     measurement2      x   x   x   x   x   x   x   x
>     measurement3        x   x   x   x   x   x   x
>     measurement4          x   x   x   x   x   x   x
> 
>     time (s) -->    1   2   3   4   5   6   7
>                     |        |
>                     '--------'
>                         |
>                       warmup
> ```
> 
> With your implementation, the measurement4 would not be completely warmed up, as we would take the measurement1.start_timestamp as the basis for the addition.
> 
> All I'm saying is that we can't use `self.start_timestamp` but most likely `max([res.start_timestamp) for res in self.generator_results + self.receiver_results + ...])`.

the diagram you created is almost correct and i aware of this situation... however the old code doesn't work correctly either due to bad assumptions and also simply crashes when the measurement hierarchy is created slightly differently

the point i am making is that the difference in start times in your diagram is IMO greatly exaggerated... when i tried this locally with 5 streams  the difference between the new calculation and the old calculation was... 0.002s which imo is simply not worth it for the code complexity. I have scheduled the tests with the large numbers of streams to see what really happens.

Where your diagram ISN'T correct is that you're writing `measurement 1...2...3...4` but this warmup warmdown is localized to a single measurement but different measurement "subparts": generator vs receiver and multiple streams... for this it should be almost irrelevant because this will be the same Iperf test module anyway...

For parallel processes the single measurement will run multiple iperf test modules but these are prepared and started in a way that should hopefully minimize delay between them... but i do want to see what we get from the scheduled beaker job.

Reply:
Yeah, I know I've exagerrated. This is solely for the purpose of flaw in the implementation.

We're considering only iperf/neper flow measurements - what if we add another where the startups are somewhat lazier?

There's (I believe) minimal complexity.

My proposal is simply:
```
        return max([res.start_timestamp for res inself.generator_results + self.receiver_results +...]) + self.warmup_duration
```

Whereas your implementation basically returns:
```
        return min([res.start_timestamp for res inself.generator_results + self.receiver_results +...]) + self.warmup_duration
```

Because `self.start_timestamp` is exactly the same except the it's `min()`: https://github.com/LNST-project/lnst/blob/de79872547e204f3cc4b91b496dfbf0fd923f60d/lnst/RecipeCommon/Perf/Measurements/Results/FlowMeasurementResults.py#L64